components/encryption/src/master_key/kms.rs

```diff
@@ -0,0 +1,306 @@
+
+use super::{metadata::MetadataKey, Backend, FileBackend, PlainTextBackend, WithMetadata};
+use crate::config::KmsConfig;
+use crate::encrypted_file::EncryptedFile;
+use crate::{Error, Result};
+
+const KMS_ENCRYPTION_KEY_NAME: &str = "kms_encryption.key";
+
+// Always use AES 256 for encrypting master key.
+const KMS_DATA_KEY_METHOD: EncryptionMethod = EncryptionMethod::Aes256Ctr;
+const AWS_KMS_DATA_KEY_SPEC: &str = "AES_256";
+const AWS_KMS_VENDOR_NAME: &[u8] = b"AWS";
+
+struct AwsKms {
+    client: KmsClient,
+    current_key_id: String,
+}
+
+impl AwsKms {
+    fn new(config: KmsConfig) -> Result<AwsKms> {
+        let http_dispatcher = HttpClient::new().unwrap();
+
+        AwsKms::with_request_dispatcher(config, http_dispatcher)
+    }
+
+    // TODO following code is almost the same as external_storage s3 client,
+    //      should be wrapped together.
+    fn with_request_dispatcher<D>(config: KmsConfig, dispatcher: D) -> Result<AwsKms>
+    where
+        D: DispatchSignedRequest + Send + Sync + 'static,
+    {
+        if config.key_id.is_empty() {
+            return Err(Error::Other(
+                "KMS key id can not be empty".to_owned().into(),
+            ));
+        }
+
+        let region = if config.endpoint.is_empty() {
+            config.region.parse::<region::Region>().map_err(|e| {
+                Error::Other(format!("invalid region format {}: {}", config.region, e).into())
+            })?
+        } else {
+            region::Region::Custom {
+                name: config.region,
+                endpoint: config.endpoint,
+            }
+        };
+
+        let client = if config.access_key.is_empty() || config.secret_access_key.is_empty() {
+            let cred_provider = DefaultCredentialsProvider::new()
+                .map_err(|e| Error::Other(format!("unable to get credentials: {}", e).into()))?;
+            KmsClient::new_with(dispatcher, cred_provider, region)
+        } else {
+            let cred_provider = StaticProvider::new(
+                config.access_key,
+                config.secret_access_key,
+                None, /* token */
+                None, /* valid_for */
+            );
+            KmsClient::new_with(dispatcher, cred_provider, region)
+        };
+
+        Ok(AwsKms {
+            client,
+            current_key_id: config.key_id,
+        })
+    }
+
+    fn decrypt(&self, runtime: &mut Runtime, ciphertext: &[u8]) -> Result<Vec<u8>> {
+        let decrypt_request = DecryptRequest {
+            ciphertext_blob: ciphertext.to_vec().into(),
+            encryption_context: None,
+            grant_tokens: None,
+            encryption_algorithm: None,
+            key_id: None,
+        };
+        let decrypt_response = retry(runtime, || {
+            self.client
+                .decrypt(decrypt_request.clone())
+                .map_err(|e| Error::Other(e.into()))
+        });
+        let plaintext = decrypt_response.plaintext.unwrap().as_ref().to_vec();
```

Comment:
verify if `key_id` from the response is the same as `current_key_id`, if not, we should regenerate a new data key, since master key is rotated. (Or shall we fail the whole `KmsBackend` and handle master key fallback logic at higher level?)

Reply:
It's supposed to fail if key_id mismatch and let's handle master rotation at a higher level.